src/bootstrap/src/core/builder.rs

```diff
@@ -72,36 +72,40 @@ impl<'a> Deref for Builder<'a> {
 }
 
 pub trait Step: 'static + Clone + Debug + PartialEq + Eq + Hash {
-    /// `PathBuf` when directories are created or to return a `Compiler` once
-    /// it's been assembled.
+    /// Result type of `Step::run`.
     type Output: Clone;
 
     /// Whether this step is run by default as part of its respective phase.
-    /// `true` here can still be overwritten by `should_run` calling `default_condition`.
+    ///
+    /// Note: Even if set to `true`, it can still be overridden (with `default_condition`) by `should_run`.
     const DEFAULT: bool = false;
 
     /// If true, then this rule should be skipped if --target was specified, but --host was not
     const ONLY_HOSTS: bool = false;
 
-    /// Primary function to execute this rule. Can call `builder.ensure()`
-    /// with other steps to run those.
+    /// Primary function to implement `Step` logic.
+    ///
+    /// This function can be triggered in two ways:
+    ///     1. Directly from bootstrap `Step` handler.
+    ///     2. Indirectly by being called from other `Step`s using `builder.ensure()`.
     ///
-    /// This gets called twice during a normal `./x.py` execution: first
-    /// with `dry_run() == true`, and then for real.
+    /// When called directly from bootstrap, this function is called twice:
+    ///     - First in "dry-run" mode to validate certain things (like cyclic Step invocations,
+    ///         directory creation, etc) super quickly.
+    ///     - Then it's called again to run the actual, very expensive process.
```

Comment:
Remark: I almost wonder if this wants to be split into `dry_run` and `run` instead of always having to remember checking `if dry_run { ... }` lol. Implementation details shared between `dry_run` and `run` can always be shared by delegating repeated parts to other methods on the `impl Step for $StepName { }` impl block.